res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+                        // are disconnected, so overwrite the whole array here instead
+                        // of assigning to this.connections[0] to avoid
+                        // Component.prototype.trigger() triggering the disconnected connection.
+                        this.connections = [engine.makeConnection(eu.group,
+                            "focused_effect",
+                            this.output.bind(this))];
+                    };
+                }
+            };
+
+            this.knobs = new ComponentContainer();
+            this.enableButtons = new ComponentContainer();
+            for (let n = 1; n <= 3; n++) {
+                this.knobs[n] = new this.EffectUnitKnob(n);
+                this.enableButtons[n] = new this.EffectEnableButton(n);
+            }
+
+            this.effectFocusButton = new Button({
+                group: this.group,
+                longPressed: false,
+                longPressTimer: 0,
+                pressedWhenParametersHidden: false,
+                previouslyFocusedEffect: 0,
+                startEffectFocusChooseMode: function() {
+                    if (colors !== undefined) {
+                        this.color = colors.focusChooseMode;
+                    }
+                    this.send(this.on);
+                    eu.focusChooseModeActive = true;
+                    eu.enableButtons.reconnectComponents(function(button) {
+                        button.startEffectFocusChooseMode();
+                    });
+                },
+                setColor: function() {
+                    if (colors !== undefined) {
+                        if (engine.getValue(this.group, "focused_effect") === 0) {
+                            this.color = colors.unfocused;
+                        } else {
+                            this.color = colors.focused;
+                        }
+                    }
+                },
+                unshift: function() {
+                    this.input = function(channel, control, value, status, _group) {
+                        const showParameters = engine.getValue(this.group, "show_parameters");
+                        if (this.isPress(channel, control, value, status)) {
+                            this.longPressTimer = engine.beginTimer(this.longPressTimeout,
+                                this.startEffectFocusChooseMode.bind(this),
+                                true);
+                            if (!showParameters) {
+                                if (!allowFocusWhenParametersHidden) {
+                                    engine.setValue(this.group, "show_parameters", 1);
+                                    // eu.onShowParametersChange will refocus the
+                                    // previously focused effect and show focus in skin
+                                }
+                                this.pressedWhenParametersHidden = true;
+                            }
+                        } else {
```

Comment:
the nesting gets very deep here.  Where possible, can you do 
```
if { 
  ... 
  return;  
 }
...
```

instead of:
```
if {
...
} else {
...
}
```

Reply:
Sure, I'm also favoring early returns, this was just not my code to begin with, but I can change it now surely.